File: 2023/p2.ml
let read_lines () =
  let ch = open_in "input_p2.txt" in
  let get_line () =
    try Some (input_line ch) with End_of_file -> None
  in 
  let rec f acc =
    match get_line () with 
    | None -> List.rev acc
    | Some line -> f (line :: acc)
  in 
  f []

type draw = {
  red : int;
  green : int;
  blue : int;
}

type game = {
  id : int;
  draws : draw list;
}

let init_draw () = { red = 0; green = 0; blue = 0 }

let line_to_game str =
  let parse_draws draws =
    let nc_strs = List.map String.trim (String.split_on_char ',' draws) in
    let ncs = List.map (fun s -> let nc = String.split_on_char ' ' s in (int_of_string (List.hd nc), List.nth nc 1)) nc_strs in
    let update_draw draw (num, col) = match col with
    | "red" -> { draw with red = num }
    | "green" -> { draw with green = num }
    | "blue" -> { draw with blue = num}
    | _ -> assert false
    in
    List.fold_left update_draw (init_draw ()) ncs
  in
  let parse_draws_line line =
    let draws = List.map String.trim (String.split_on_char ';' line) in
    List.map parse_draws draws
  in
  let re = Str.regexp "Game \\([0-9]+\\):\\(.*$\\)" in 
  let _ = Str.string_match re str 0 in
  let id_str = Str.matched_group 1 str in
  let draws_str = Str.matched_group 2 str in
  {
    id = int_of_string id_str;
    draws = parse_draws_line draws_str;
  }

let is_possible game =
  let is_possible_draw draw =
    draw.red <= 12 && draw.green <= 13 && draw.blue <= 14
  in
  List.for_all is_possible_draw game.draws

let power game =
  let max_draw a b =
    {
      red = max a.red b.red;
      green = max a.green b.green;
      blue = max a.blue b.blue;
    }
  in
  let { red; green; blue } = List.fold_left max_draw (init_draw ()) game.draws in
  red * green * blue

let () =
  let lines = read_lines() in
  let games = List.map line_to_game lines in
  let ans1 = List.fold_left (fun acc game -> if is_possible game then game.id + acc else acc) 0 games in 
  let ans2 = List.fold_left (fun acc game -> acc + power game) 0 games in
  Printf.printf "Answer to part 1 is %d\nAnswer to part 2 is %d\n" ans1 ans2

File: 2023/p2.cpp
#include <bits/stdc++.h>

using namespace std;

vector<string> ReadFile(const string& filename) {
    ifstream is(filename);

    vector<string> content;
    string s;
    while (getline(is, s)) {
        content.push_back(s);
    }

    return content;
}

struct Draw {
    int red;
    int green;
    int blue;

    Draw() : red(0), green(0), blue(0) {}
};

class Game {
private:
    int id_;
    vector<Draw> draws_;

public:
    Game(int id) : id_(id) {}

    int GetID() const {
        return id_;
    }

    vector<Draw> GetDraws() const {
        return draws_;
    }

    void AddDraw(const Draw& draw) {
        draws_.push_back(draw);
    }

    int Power() const {
        int red = 0;
        int green = 0;
        int blue = 0;
        for (const Draw& draw : draws_) {
            red = max(red, draw.red);
            green = max(green, draw.green);
            blue = max(blue, draw.blue);
        }
        return red * green * blue;
    }
};

Game ParseGame(const string& line) {
    int i = line.find(' '); i++;
    int j = line.find(':', i);
    int id = stoi(line.substr(i, j - i));
    Game game(id);

    for (i = j + 2; i < line.size();) {
        Draw draw;
        while (true) {
            for (j = i + 1; line[j] != ' '; j++) {}
            int n = stoi(line.substr(i, j - i));
            i = j + 1;

            for (j = i + 1; j < line.size() && isalpha(line[j]); j++) {}
            string colour = line.substr(i, j - i);
            i = j + 2;

            if (colour == "red") draw.red = n;
            else if (colour == "green") draw.green = n;
            else draw.blue = n;

            if (j == line.size() || line[j] == ';') break;
        }
        game.AddDraw(draw);
    }

    return game;
}

vector<Game> ParseGames(const vector<string>& lines) {
    vector<Game> games;
    for (const string& line : lines) {
        games.push_back(ParseGame(line));
    }
    return games;
}

bool Possible(const Game& game) {
    for (const Draw& draw : game.GetDraws()) {
        if (draw.red > 12 || draw.green > 13 || draw.blue > 14) return false;
    }
    return true;
}

int main() {
    vector<string> lines = ReadFile("input_p2.txt");
    vector<Game> games = ParseGames(lines);

    int ans1 = 0;
    for (const Game& game : games) {
        if (Possible(game)) {
            ans1 += game.GetID();
        }
    }

    int ans2 = 0;
    for (const Game& game : games) {
        ans2 += game.Power();
    }

    printf("Answer to part 1 is %d\nAnswer to part 2 is %d\n", ans1, ans2);

    return 0;
}